### PR TITLE
Enhance methodswith

### PR DESCRIPTION
**Pull request by timholy**
See #33866. This pretty much does everything I need it to currently. It does not tackle the [code comment](https://github.com/JuliaLang/julia/blob/4982b36cc098e98b945b10ee365a88e66d8969ee/stdlib/InteractiveUtils/src/InteractiveUtils.jl#L118), though it does essentially work around it via subtyping.

### CODE REVIEW

File: stdlib/InteractiveUtils/test/runtests.jl [hunk 1]
```diff
@@ -12,6 +12,57 @@ struct Type4Union end
 func4union(::Union{Type4Union,Int}) = ()
 @test !isempty(methodswith(Type4Union, @__MODULE__))
 
+module MWM
+export f, SubModule
+struct C end
+abstract type Super end
+struct Sub <: Super end
+abstract type SuperP{A<:AbstractArray} end
+abstract type SubP{A<:AbstractVector} <: SuperP{A} end
```
Comment (timholy) [inline stdlib/InteractiveUtils/test/runtests.jl]: Sadly, had I written `SubP{A} <: SuperP{A}` then the test would have failed. Presumably this is just #6383, but if there is a nice workaround I'd be happy to incorporate it. I am sure there's a lot of code out there that doesn't re-impose the same parameter bounds.

**Comment by timholy** [comment]
Do we want this for Julia 1.4? There seem to be a number of supportive reactions but it would be nice if this got some code review and a merge decision.

**Comment by StefanKarpinski** [comment]
I'm fine with it. @JeffBezanson? Given that this is an interactive tool, I'm not overly concerned.

**Comment by timholy** [comment]
I can achieve everything I need with `MethodAnalysis`, so let's close this.

**Pull request closed without merging**